pkg/tsdb/opentsdb/opentsdb.go

```diff
@@ -3,54 +3,114 @@ package opentsdb
 	"strconv"
 	"strings"
 	"time"
 
-	"golang.org/x/net/context/ctxhttp"
-
 	"encoding/json"
-	"io/ioutil"
-	"net/http"
-	"net/url"
 
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/datasource"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/instancemgmt"
 	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/grafana/grafana/pkg/components/simplejson"
 	"github.com/grafana/grafana/pkg/infra/httpclient"
 	"github.com/grafana/grafana/pkg/infra/log"
-	"github.com/grafana/grafana/pkg/models"
-	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/backendplugin"
+	"github.com/grafana/grafana/pkg/plugins/backendplugin/coreplugin"
+	"github.com/grafana/grafana/pkg/registry"
 	"github.com/grafana/grafana/pkg/setting"
+	"golang.org/x/net/context/ctxhttp"
 )
 
-type OpenTsdbExecutor struct {
-	httpClientProvider httpclient.Provider
+type Service struct {
+	HTTPClientProvider   httpclient.Provider   `inject:""`
+	Cfg                  *setting.Cfg          `inject:""`
+	BackendPluginManager backendplugin.Manager `inject:""`
+
+	im instancemgmt.InstanceManager
+}
+
+type datasourceInfo struct {
+	HTTPClient        *http.Client
+	Url               string `json:"url"`
+	BasicAuth         bool   `json:"basicAuth"`
+	BasicAuthUser     string `json:"basicAuthUser"`
+	BasicAuthPassword string `json:"basicAuthPassword"`
+}
+
+type DsAccess string
+
+func init() {
+	registry.Register(&registry.Descriptor{Instance: &Service{}})
+}
+
+func (s *Service) Init() error {
+	s.im = datasource.NewInstanceManager(NewInstanceSettings(s.HTTPClientProvider))
+
+	factory := coreplugin.New(backend.ServeOpts{
+		QueryDataHandler: s,
+	})
+
+	if err := s.BackendPluginManager.Register("opentsdb", factory); err != nil {
+		plog.Error("Failed to register plugin", "error", err)
+	}
+
+	return nil
 }
 
-//nolint: staticcheck // plugins.DataPlugin deprecated
-func New(httpClientProvider httpclient.Provider) func(*models.DataSource) (plugins.DataPlugin, error) {
-	//nolint: staticcheck // plugins.DataPlugin deprecated
-	return func(*models.DataSource) (plugins.DataPlugin, error) {
-		return &OpenTsdbExecutor{
-			httpClientProvider: httpClientProvider,
-		}, nil
+func NewInstanceSettings(httpClientProvider httpclient.Provider) datasource.InstanceFactoryFunc {
+	return func(settings backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
+		opts, err := settings.HTTPClientOptions()
+		if err != nil {
+			return nil, err
+		}
+
+		client, err := httpClientProvider.New(opts)
+		if err != nil {
+			return nil, err
+		}
+
+		jsonData := datasourceInfo{}
+		err = json.Unmarshal(settings.JSONData, &jsonData)
+		if err != nil {
+			return nil, fmt.Errorf("error reading settings: %w", err)
+		}
+		basicAuthPassword := settings.DecryptedSecureJSONData["basicAuthPassword"]
+		if basicAuthPassword == "" {
+			basicAuthPassword = jsonData.BasicAuthPassword
+		}
+
+		model := datasourceInfo{
```

Comment:
Wonder if this should be a pointer instead, e.g. a pointer returned from this function? In `getDSInfo` you cast to non-pointer, but return a pointer of the value - with returning a pointer here instead you should be able to cast directly to pointer in `getDSInfo`.